rfc/0003-policy-signing.md

```diff
@@ -0,0 +1,440 @@
+`https://github.com/alice/fancy-policy/.github/workflows/cd.yml@refs/heads/main` | ✅ | The policy is coming from `https://github.com/alice/fancy-policy/`
+`https://github.com/alice/another-policy/.github/workflows/cd.yml@refs/heads/main` | ❌ | The policy is coming from a repository owned by Alice, but not trusted by the configuration
+
+Given the previous configuration, we would also trust all the policies that have been
+signed with the `john@acme.org` email, as long as the identity provider is the
+GitHub website (this is the identity provider used by "human-driven" keyless signatures
+relying on the GitHub identity provider).
+
+**Note well:** `issuers` defines a per-issuer allow list. That means the policies
+**do not have** to be signed by all the subjects mentioned inside of the list.
+They must all be signed by at least one of these issuers.
+
+### Annotations
+
+This is an entirely optional section that applies both to key and keyless
+signatures.
+
+Cosign allows the signer to specify custom key-value attributes to embedded
+into the signature.
+
+```yaml
+annotations:
+  env: prod
+  foo: bar
+```
+
+Given the previous configuration, all the policies must include an annotation
+with key `env` and value `prod`, plus another one with key `foo` and value `bar`.
+
+## Configuration scenarios
+
+As stated above the verification settings are made by three sections:
+
+  * `public_keys`
+  * `issuers`
+  * `annotations`
+
+The following table describes how verification happens when these fields
+are specified (✅) or left empty (❌).
+
+| `public_keys` | `issuers` | annotations | Behaviour |
+|-------------|-----------|-------------|-----------|
+| ✅ | ❌ | ❌ | policies must be signed by at least <threshold> keys |
+| ✅ | ✅ | ❌ | policies must be signed by at least <threshold> keys, **plus** they must be signed by one of the issuers |
+| ❌ | ✅ | ❌ | policies must be signed by one of the issuers |
+| ❌ | ✅ | ✅ | policies must be signed by one of the issuers, **plus** they must satisfy all the annotations |
+| ✅ | ❌ | ✅ | policies must be signed by at least <threshold> keys, **plus** they must satisfy all the annotations |
+| ✅ | ✅ | ✅ | policies must be signed by at least <threshold> keys, plus they must be signed by one of the issuers, **plus** they must satisfy all the annotations |
+| ❌ | ❌ | ✅ | this configuration is wrong, it should result in an error |
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+> Why should we **not** do this?
+>
+>   * obscure corner cases
+>   * will it impact performance?
+>   * what other parts of the product will be affected?
+>   * will the solution be hard to maintain in the future?
+
+Policy signature verification is going to be an opt-in feature.
+
+There's no reason to not implement this feature.
+
+# Alternatives
+[alternatives]: #alternatives
+
+> - What other designs/options have been considered?
+> - What is the impact of not doing this?
+
+## Specify per-policy signature settings
```

Comment:
I think having `annotations` in a way that they can be set in `ClusterAdmissionPolicy` is interesting.

A PolicyServer `ps1` might specify a "baseline" of `annotations`: `["annotation1", "annotation2"]`.

If a `ClusterAdmissionPolicy` specifies specific policy annotations `["extra1", "extra2"]`, then the policy must meet `["annotation1", "annotation2", "extra1", "extra2"] for it to be schedulable to `ps1`. Otherwise, it can't.

Is not an "alternative" to the main RFC, just a way to make it more expressive and configurable per-policy.

Reply:
To me, the RBAC means that it's only useful for cluster admins, which diminishes returns.

Reply:
Not sure what do you mean with RBAC. Can you elaborate on that? My comment is not applicable to RBAC, but to ensure that labels can be provided per policy server and per policies. Then, a policy must have been signed with all the key=value required both in the policy-server where it's to be scheduled, and on the policy CR itself.

Reply:
In my opinion the shortcomings already listed for this in the RFC make it ugly.

Also, if we have per-policy annotations/signatures, it seems only useful for users that can deploy their own ClusterAdmissionPolicies, but don't have access to the PolicyServer. 
I would be happier using the verification config only in PolicyServers, and providing NamespacedPolicyServers. 
If what we want is a tiered system of verification config, I would prefer allowing PolicyServers to read a list of Configmaps, and some of them from a different Namespace. That way one can use the ns permissions.

Reply:
Let's ignore my comment for now. I was aiming at also having the same behaviors when we introduce the `AdmissionPolicy` namespaced resource. But I have to admit I don't have a strong story to show now, just a hunch that it would be useful to allow. It's fine to ignore it right now.